### PR TITLE
L043 refinement: Add autofix for common use of CASE to fill NULL values.

### LINKED ISSUE #1797: New rule: COALESCE instead of CASE/IF when possible
In all dialects that have `COALESCE` code such as `CASE WHEN num_cats IS NULL THEN 0 ELSE num_cats END` are bad practices because they are very long and need to be simplified as `COALESCE(num_cats, 0)` which is much shorter and a lot more readable.

### PR DESCRIPTION
**Pull request by jpy-git**
### Brief summary of the change made

This PR enhances L043 to fix #1797. I have also tidied up the docstring and added the corresponding unit tests.

### Are there any other side effects of this change that we should be aware of?
L043 should now flag and fix more cases of unnecessary use of CASE. 

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by jpy-git** [comment]
@barrywhart Ready for review 😄

File: src/sqlfluff/rules/L043.py [hunk 1]
```diff
@@ -1,154 +1,261 @@
...
-                            KeywordSegment("not"),
-                            WhitespaceSegment(),
-                        ]
-                    )
-                # Add coalesce and parenthesis
-                edits.extend(
+                coalesce_arg_1 = condition_expression
+                coalesce_arg_2 = KeywordSegment("false")
+                coalesce_arg_1_idx = condition_expression_idx
+                preceding_not = True if then_expression.raw_upper == "FALSE" else False
```
Comment (barrywhart) [inline src/sqlfluff/rules/L043.py:187]:
```suggestion
                preceding_not = (then_expression.raw_upper == "FALSE")
```

**Review by barrywhart — APPROVED** [review]
Looks good to me.

This rule has always been one of the scarier ones to me, because I worry about the risk of "false positives", like if it triggers for a corner case we didn't consider. For example, I think an early version of this rule assumed (incorrectly) that if a case statement used booleans and an else, that one boolean would be false and the other true. Can you think of any potential "gotchas" with these changes, and perhaps we should add more tests to catch them?

File: src/sqlfluff/rules/L043.py (hunk 1 above)
Comment (jpy-git) [reply src/sqlfluff/rules/L043.py]: @barrywhart 🤦 wow I definitely had a brain lapse there, thanks!

**Comment by jpy-git** [comment]
@barrywhart so I added some checks to make sure we only operate on CASE statements with a single WHEN and also added a check that the IS NULL condition is the only condition. Iet me make some extra unit tests for cases we think should pass this lint unaffected for extra peace of mind 👍

**Comment by jpy-git** [comment]
@barrywhart ive added additional unit test to validate my checks in the code so this should be good to go 🚀

**Pull request merged**